WIP of gpio pin ctrl

annerajb opened the pull request:
## Summary

## Impact

## Testing

annerajb committed:
WIP of gpio pin ctrl

annerajb commented on arch/arm64/src/rk3399/mx8mp_iomuxc.c:
```diff
@@ -0,0 +1,81 @@
+ * Name: mx8mp_iomuxc_config
+ ****************************************************************************/
+
+void mx8mp_iomuxc_config(uint32_t mux_register,
```
@lupyuen
Thinking about adding a new argument here to indicate which pin on the iomux register of that bank is intended to be configured between the mux_mode

annerajb replied:
Added it as part of the define arguments

annerajb commented on arch/arm64/src/rk3399/hardware/mx8mp_gpio.h:
```diff
@@ -0,0 +1,87 @@
+/* GPIO Register Addresses **************************************************/
+//if gpio2 add + 6 0000 due to different base address
+//TODO: should we do this as a inline function 
+#define GPIO_BASE_ADDR(n)            ((2 >= n ) ? RK3399_GPIO0_ADDR : RK3399_GPIO2_ADDR)
```
@lupyuen
This is the best Idea I got for being able to switch the base address between the PMUGRF and GRF since their gpio have different base addresses. 
I have todo this also on the iomux file but was wondering what suggestions you had that could be better.

lupyuen commented:
The RK3399 GPIO Fields look surprisingly similar to this: 

- [arch/arm/src/phy62xx/mcu_phy_bumbee.h](https://github.com/apache/nuttx/blob/master/arch/arm/src/phy62xx/mcu_phy_bumbee.h#L389-L415)

- [arch/arm/src/phy62xx/gpio.c](https://github.com/apache/nuttx/blob/master/arch/arm/src/phy62xx/gpio.c)

Maybe we could adapt the code from above? It might be hard to merge RK3399 and PHY62xx into the same source file, since its Arm64 vs Arm32. So we might have to physically copy the source.

(It's the same for Allwinner A64: I copied the NuttX GPIO Driver from Allwinner A1x)

Might be interesting to have a chat with the creator of the above source file, to understand what's PHY62xx and whether it shares the same hardware as RK3399. Thanks!

annerajb committed:
working iomux prototype / concept
wip gpio with selectable bank / gpio addr skeleton

annerajb commented on arch/arm64/src/rk3399/mx8mp_gpio.c:
```diff
@@ -0,0 +1,539 @@
+  irqstate_t flags;
+
+  flags = enter_critical_section();
+  r = hal_gpio_read(m_gpiobanks[port].bankAddr,pin);
```
@lupyuen came up with this to allow to use the pinset to specify which of the 5 gpio to use and use the structure to contain additional data such as if the grf is on the pmugrf  vs regular grf.

Need to finish implementing the hal init and hal write. 
Probably first version won't implement interrupts.

lupyuen replied:
Yep looks OK thanks!

annerajb commented on arch/arm64/src/rk3399/hardware/mx8mp_pinmux.h:
```diff
@@ -0,0 +1,197 @@
+#define IOMUXC_GRF_GPIO4C_IO07_EDP_HOTPLUG
+// GPIO 4D
+
+#define IOMUXC_GRF_GPIO4D_IO00_GPIO          GRF_GPIO4D_IOMUX_ADDR, HIWORD_UPDATE(0b0000,0b1111,0), 0x00000000, 0x0, 0x00000000
```
Need to rename hiword update but this is how it would look like to specify a specific grf usage / config

lupyuen replied:
Yep sure

annerajb committed:
added more registers and add iomux call from bringup
still WIP since it includes the GPIO and I have to finish more registers on rk3399_pinmux.h
